attic/multibody/multibody_solvers.cc

```diff
@@ -0,0 +1,835 @@
+      num_positions_(tree.get_num_positions()),
+      num_velocities_(tree.get_num_velocities()),
+      num_efforts_(tree.get_num_actuators()),
+      num_position_forces_(tree.getNumPositionConstraints()),
+      num_forces_(tree.getNumPositionConstraints() +
+                  contact_info.num_contacts * 3) {
+  // ContactToolkit pointer using the ContactInfo object.
+  contact_toolkit_ =
+      make_unique<ContactToolkit<AutoDiffXd>>(tree, contact_info);
+}
+
+void FixedPointConstraint::DoEval(
+    const Eigen::Ref<const Eigen::VectorXd>& q_u_l, Eigen::VectorXd* y) const {
+  AutoDiffVecXd y_t;
+  Eval(initializeAutoDiff(q_u_l), &y_t);
+  *y = autoDiffToValueMatrix(y_t);
+}
+
+void FixedPointConstraint::DoEval(
+    const Eigen::Ref<const drake::AutoDiffVecXd>& q_u_l,
+    drake::AutoDiffVecXd* y) const {
+  // Verifying the size of the input vector
+  DRAKE_DEMAND(q_u_l.size() == num_positions_ + num_efforts_ + num_forces_);
+
+  // Extracting the components
+  const AutoDiffVecXd q = q_u_l.head(num_positions_);
+  const AutoDiffVecXd u = q_u_l.segment(num_positions_, num_efforts_);
+  const AutoDiffVecXd lambda = q_u_l.tail(num_forces_);
+
+  // the velocities are assumed to be zero.
+  const AutoDiffVecXd v =
+      VectorXd::Zero(num_velocities_).template cast<AutoDiffXd>();
+  AutoDiffVecXd x(num_positions_ + num_velocities_);
+  x << q, v;
+
+  // The constraint is set up using MVDot as it is more stable than computing
+  // xdot and constraining it to be zero.
+  *y = contact_toolkit_->CalcMVDot(x, u, lambda);
+}
+
+void FixedPointConstraint::DoEval(
+    const Eigen::Ref<const drake::VectorX<drake::symbolic::Variable>>& q_u_l,
+    drake::VectorX<drake::symbolic::Expression>* y) const {
+  throw logic_error(
+      "FixedPointConstraint does not support symbolic evaluation.");
+}
+
+PositionSolver::PositionSolver(const RigidBodyTree<double>& tree)
+    : tree_(tree), prog_(make_shared<MathematicalProgram>()) {
+  // This constructor does not setup the quadratic cost.
+
+  // Initializing the variable
+  q_ = prog_->NewContinuousVariables(tree_.get_num_positions(), "q");
+
+  // Solver setup
+  prog_->SetSolverOption(SnoptSolver::id(), "Print file", filename_);
+  prog_->SetSolverOption(SnoptSolver::id(), "Major feasibility tolerance",
+                         major_tolerance_);
+  prog_->SetSolverOption(SnoptSolver::id(), "Minor feasibility tolerance",
+                         minor_tolerance_);
+
+  auto position_constraint = make_shared<PositionConstraint>(tree_);
+
+  prog_->AddConstraint(position_constraint, q_);
+}
+
+PositionSolver::PositionSolver(const RigidBodyTree<double>& tree,
+                               VectorXd q_desired, MatrixXd Q)
+    : tree_(tree), prog_(make_shared<MathematicalProgram>()) {
+  // Checking the validity of q_desired
+  DRAKE_DEMAND(q_desired.size() == tree_.get_num_positions());
+
+  // If the user does not provide Q, default it to identity
+  if (Q.rows() == 0 && Q.cols() == 0) {
```

Comment:
We shouldn't be passing in arguments that we know are wrong. I see why you did this, because you can't set a default parameter without knowing the dimension, but this isn't right. Here's what I recommend for the constructors:

```
 // default constructor, all other constructors delegate here
PositionSolver(tree) {
// basic program setup
}

PositionSolver(tree, q, Q) : PositionSolver(tree) {
  //Calls default constructor then adds cost
  AddProgramCost(q,Q);
}

PositionSolver(tree,q) : PositionSolver(tree,q,Identity(q.rows())) {}

Reply:
This way, there's no duplicated code between constructors, and no strange if statements. Also, `q` and `Q` can and should be made `const` this way.

Reply:
I see. The matrix check did seem to be a hacky way of doing it. I'll update it.

Reply:
I guess I can also remove the default Q value from `AddProgramCost`. If the user calls it specifically, it's probably to change it from something other than identity.

Reply:
dd09d84 implements this. 
* The default values are initialized through the constructor.
* Constructor code is reused.
* Default `Q` and `U` values from `AddProgramCost` have been removed.